platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/tipset/TipsetScoreCalculator.java

```diff
@@ -0,0 +1,254 @@
+    }
+
+    /**
+     * Get the current tipset snapshot.
+     *
+     * @return the current tipset snapshot
+     */
+    public @NonNull Tipset getSnapshot() {
+        return snapshot;
+    }
+
+    /**
+     * Add an event created by this node and compute the increase in tipset score. Higher score changes mean that this
+     * event caused consensus to advance more. A score change of 0 means that this event did not advance consensus. A
+     * score change close to the total weight means that this event did a very good job at advancing consensus. It's
+     * impossible to get a perfect score, since the weight of advancing self events is not included. The maximum score
+     * an event can achieve is equal to the sum of all weights minus the sum of this node's weight.
+     *
+     * @param event the event that is being added
+     * @return the change in the tipset advancement score
+     */
+    public long addEventAndGetAdvancementScore(@NonNull final EventDescriptor event) {
+        Objects.requireNonNull(event);
+        if (!event.getCreator().equals(selfId)) {
+            throw new IllegalArgumentException("event creator must be the same as the window ID");
+        }
+
+        final Tipset eventTipset = tipsetBuilder.getTipset(event);
+        if (eventTipset == null) {
+            throw new IllegalArgumentException("event " + event + " is not in the tipset tracker");
+        }
+
+        final long score = snapshot.getWeightedAdvancementCount(selfId, eventTipset);
+        if (score > maximumPossibleScore) {
+            throw new IllegalStateException(
+                    "score " + score + " is greater than the maximum possible score " + maximumPossibleScore);
+        }
+
+        final long scoreImprovement = score - previousScore;
+
+        if (isSuperMajority(score + selfWeight, totalWeight)) {
+            snapshot = eventTipset;
+            snapshotHistory.add(snapshot);
+            if (snapshotHistory.size() > snapshotHistorySize) {
+                snapshotHistory.remove();
+            }
+            previousScore = 0;
+        } else {
+            previousScore = score;
+        }
+
+        return scoreImprovement;
+    }
+
+    /**
+     * Figure out what advancement score we would get if we created an event with a given list of parents.
+     *
+     * @param parents the proposed parents of an event
+     * @return the advancement score we would get by creating an event with the given parents
+     */
+    public long getTheoreticalAdvancementScore(@NonNull final List<EventDescriptor> parents) {
+        if (parents.isEmpty()) {
+            return 0;
+        }
+
+        final List<Tipset> parentTipsets = new ArrayList<>(parents.size());
+        for (final EventDescriptor parent : parents) {
+            parentTipsets.add(tipsetBuilder.getTipset(parent));
+        }
+
+        // Don't bother advancing the self generation, since self advancement doesn't contribute to tipset score.
```

Comment:
Let me know if you'd like me to expand the documentation here. Here is the long explanation.

This method is called when we are thinking about creating a new event. We would like to compare the tipset of this theoretical event to the current snapshot.

In order to construct the theoretical tipset, we need to merge the parent tipsets and advance the self generation.
However, since the self generation is ignored when generating this score, we can take a shortcut and not bother updating it. The advancement score of the tipset just from the merged parents will be the same as the advancement score that includes the proper self generation.

Since we throw away this tipset and just return a score, there is no problem with skipping a little extra work. This comment was intended to explain why we are not actually constructing the true tipset of the theoretical child.

Reply:
Gotcha. I think if you modified the comment slightly, it would be more clear to me (and hopefully others, too)

"Don't bother advancing the self generation *in this theoretical tipset*, since self advancement..."

Reply:
changed to suggested wording

```
        // Don't bother advancing the self generation in this theoretical tipset,
        // since self advancement doesn't contribute to tipset score.
        final Tipset newTipset = Tipset.merge(parentTipsets);
```